ZIR-285: add writeup of the procedure for creating new bigint2 precompiles (#117)
risc0/zirgen: zirgen/Dialect/BigInt/Precompiles.md

# Creating a precompile using the BigInt2 system

Creating a new accelerator using BigInt2 requires a series of related changes across files in two repositories. This is the process you should follow if you want to create a new bigint2 precompile and make it available within the zkVM.

To keep the demo simple, let's imagine that we want to create a trivial accelerator which sums two 128-bit numbers, which we'll call "add128".

> First, create files under `zirgen/circuit/bigint/` which will hold the accelerator builder function. We might call them `add128.h` and `add128.cpp`, declaring and then defining a function like this:

Reviewer: ```suggestion
First, create files under [`zirgen/circuit/bigint/`] which will hold the accelerator builder function. We might call them `add128.h` and `add128.cpp`, declaring and then defining a function like this:
```

To make use of the new function, we must add it as an option of the `bigint2c` compiler program, which generates the precompiled blobs we'll need to load into the zkVM. Inside `zirgen/circuit/bigint/bigint2c.cpp`, we'll add a new command line option:

```
enum class Program {
  ModPow65537,
  EC_Double,
  EC_Add,
  Add128
};
```

Make this option available by extending the arg parser's declaration:
```
   
               clEnumValN(Program::Add128, "add128", "Add128")),
```

Later, in the main function, we must handle the new option by invoking our new generator function:

```
case Program::Add128:
  zirgen::BigInt::genAdd128(builder, loc);
  break;
```

The `bigint2c` program can now compile our addition function into a "bibc" format blob. In order to use this blob from the `risc0` repo, we must add it to the bootstrap process, which means we must integrate blob-generation into the zirgen circuit build process. This happens in `zirgen/circuit/bigint/BUILD.bazel`, where we must create a new genrule for the new blob file:

```
genrule(
    name = "add128",
    outs = ["add128.blob"],
    exec_tools = [":bigint2c"],

>     cmd = "$(location //zirgen/circuit/bigint:bigint2c) --program=add128 > $(OUTS)"

Reviewer 2: Program generation takes a mandatory bitwidth after https://github.com/risc0/zirgen/pull/107
```suggestion
    cmd = "$(location //zirgen/circuit/bigint:bigint2c) --program=add128 --bitwidth 128 > $(OUTS)"
```

)
```

Next we must add the new target name to the list of blob files to generate:

```
BLOBS = [
    "modpow65537_4096",
    "ec_double_256",
    "ec_add_256",
    "add128",
]
```

We can now `bazelisk build //zirgen/circuit/bigint/...` to produce our new blob along with the others. In order to make this available to the `risc0` world, we must add the new blob to the bootstrap process. In `zirgen/bootstrap/src/main.rs`, in the `bigint2` function, after a series of similar lines, we must add these lines to copy the new blob over:

```
        let add128_path = risc0_root.join("bigint2/src/add128");
        self.copy_file(&src_path, &add128_path, "add128.blob");
```

> We may now bootstrap the risc0 repository, including the new blob file, like this:

Reviewer 2: This command would not work on my system (as my code is in `~/code/risc0`), so maybe we should provide a hint to double check the path?
```suggestion
We may now bootstrap the risc0 repository, including the new blob file, like this (you may need to adjust the path):
```

```
cargo bootstrap bigint2 --output=`$HOME`/risc0/
```

To invoke the new addition program, we move over to the `risc0` repository. Here we will create a new module, akin to `bigint2/src/ec` or `bigint2/src/rsa`, containing a new `bigint2/src/add128/mod.rs` file. Its contents should look like this:

```
use include_bytes_aligned::include_bytes_aligned;
use crate::ffi::sys_bigint2_3;

const BLOB: &[u8] = include_bytes_aligned!(4, "add128.blob");

> pub fn add128(lhs: &[u32; 4], rhs: &[u32; 4], result: &mut [u32; 4]) {

Reviewer 2: This is not correct for this op, u128 + u128 = u256.

I think ideally we'd also create a section about checking the output bitwidth when creating this code.

```suggestion
pub fn add128(lhs: &[u32; 4], rhs: &[u32; 4], result: &mut [u32; 8]) {
```

    unsafe {
        sys_bigint2_3(
            BLOB.as_ptr(),
            lhs.as_ptr() as *const u32,
            rhs.as_ptr() as *const u32,
            result.as_mut_ptr() as *mut u32,
        );
    }
}
```

> One can now import the `add128` function to perform simple addition through the bigint2 accelerator.

Reviewer: Can we point users to where they should look to see this pattern?

Reviewer 2: Perhaps our [RSA patch](https://github.com/RustCrypto/RSA/compare/master...risc0:RustCrypto-RSA:v0.9.6-risczero.0) is the simplest "real" example?

Alternatively, we have [bigint test methods](https://github.com/risc0/risc0/blob/main/risc0/bigint2/methods/guest/src/bin/rsa.rs) that might also serve as good examples; less "real" (although they are used for tests) which might make them more example-like

Reviewer: hmm I can make a suggestion here but it sounded like you were planning to open a PR to extend this section?

Reviewer: agreed that the RSA patch is the simplest example to point to. seems like the missing steps are something like: 
- build a program wrapper
- patch a crate
- now you're ready to use the precompile in your

Reviewer: we don't need to make this super polished -- mostly just need it to be complete enough that it's clear where the current pain points are

Reviewer 2: > hmm I can make a suggestion here but it sounded like you were planning to open a PR to extend this section?

Ha, crazy timing, I just put up https://github.com/risc0/zirgen/pull/121

Reviewer: this PR comes to mind as something that might make sense to include... 
https://github.com/risc0/risc0/pull/2505

> One can now import the `add128` function to perform simple addition through the bigint2 accelerator.

Reviewer: ```suggestion
One can now import the `add128` function to perform simple addition through the bigint2 accelerator.

[`zirgen/circuit/bigint/`]: https://github.com/risc0/zirgen/tree/main/zirgen/circuit/bigint
```

# Creating a precompile using the BigInt2 system

> Creating a new accelerator using BigInt2 requires a series of related changes across files in two repositories. This is the process you should follow if you want to create a new bigint2 precompile and make it available within the zkVM.

Reviewer: ```suggestion
Creating a new precompile using BigInt2 requires a series of related changes across files in two repositories. This is the process you should follow if you want to create a new bigint2 precompile and make it available within the zkVM.
```

> To keep the demo simple, let's imagine that we want to create a trivial accelerator which sums two 128-bit numbers, which we'll call "add128".

Reviewer: ```suggestion
To keep the demo simple, let's imagine that we want to create a trivial precompile which sums two 128-bit numbers, which we'll call "add128".
```

> First, create files under [`zirgen/circuit/bigint/`] which will hold the accelerator builder function. We might call them `add128.h` and `add128.cpp`, declaring and then defining a function like this:

Reviewer: ```suggestion
## Write a bigint program
First, create files under [`zirgen/circuit/bigint/`] which will hold the accelerator builder function. We might call them `add128.h` and `add128.cpp`, declaring and then defining a function like this:
```

Reviewer: > accelerator builder function
Maybe we can just say "bigint program" here. What do you think @tzerrell?

Reviewer 2: I'm personally leaning toward "bigint program" at the moment when I discuss things myself. I might also change accelerator to precompile
```suggestion
## Write a bigint program
First, create files under [`zirgen/circuit/bigint/`] which will hold the precompile builder function. We might call them `add128.h` and `add128.cpp`, declaring and then defining a function like this:
```

```

> void genAdd128(mlir::OpBuilder& builder, mlir::Location loc) {

Reviewer: @tzerrell i wonder whether there's a reference we could add here that would enable ppl to teach themselves how to write these programs...

- somewhere in the MLIR docs that would be helpful orientation? 
- a code comment to clarify some of the details of the dialect? 
- somewhere in the overview md doc in the `Dialect/` directory?

Reviewer 2: I think this would be a substantial additional bit of work. Nothing _too_ hard, but I don't think that the mainline MLIR tutorials really cover OpBuilder, and its [reference docs](https://mlir.llvm.org/doxygen/classmlir_1_1OpBuilder.html) would not be helpful here. That means we'd have to write our own tutorial -- it's more than just a comment or even a paragraph.

This might be worth doing, and it's probably not that big a project, but it's its own thing.

Reviewer: let's not build a tutorial right now -- i thinking more along the lines of "go look at the RSA example and the underlying struct definitions and good luck"

    auto lhs = builder.create<BigInt::LoadOp>(loc, 128, 11, 0);
    auto rhs = builder.create<BigInt::LoadOp>(loc, 128, 12, 0);
    auto sum = builder.create<BigInt::AddOp>(loc, lhs, rhs);  
    builder.create<BigInt::StoreOp>(loc, sum, 13, 0);
}
```

We might also like to create a test function here, but we'll gloss over that for now.

> To make use of the new function, we must add it as an option of the `bigint2c` compiler program, which generates the precompiled blobs we'll need to load into the zkVM. Inside `zirgen/circuit/bigint/bigint2c.cpp`, we'll add a new command line option:

Reviewer: ```suggestion
## Compile your bigint program to a bigint blob using bigint2c
To make use of the new function, we must add it as an option of the `bigint2c` compiler program, which generates the precompiled blobs we'll need to load into the zkVM. Inside `zirgen/circuit/bigint/bigint2c.cpp`, we'll add a new command line option:
```

```
enum class Program {
  ModPow65537,
  EC_Double,
  EC_Add,
  Add128
};
```

Make this option available by extending the arg parser's declaration:
```
   
               clEnumValN(Program::Add128, "add128", "Add128")),
```

Later, in the main function, we must handle the new option by invoking our new generator function:

```
case Program::Add128:
  zirgen::BigInt::genAdd128(builder, loc);
  break;
```

> 

Reviewer 2: I think bootstrapping is a separate section from bigint blob compilation; that section describes how to add the ability to compile a single precompile, whereas the bootstrapping section is about adding the new precompile to the list of all precompiles to be built when by our bootstrapping command (i.e., our "compile all our precompiles and copy them to the `risc0` repo" command)

```suggestion

## Bootstrapping your precompile into the `risc0` repo

```

```
genrule(
    name = "add128",
    outs = ["add128.blob"],
    exec_tools = [":bigint2c"],
    cmd = "$(location //zirgen/circuit/bigint:bigint2c) --program=add128 --bitwidth 128 > $(OUTS)"
)
```

Next we must add the new target name to the list of blob files to generate:

```
BLOBS = [
    "modpow65537_4096",
    "ec_double_256",
    "ec_add_256",
    "add128",
]
```

We can now `bazelisk build //zirgen/circuit/bigint/...` to produce our new blob along with the others. In order to make this available to the `risc0` world, we must add the new blob to the bootstrap process. In `zirgen/bootstrap/src/main.rs`, in the `bigint2` function, after a series of similar lines, we must add these lines to copy the new blob over:

```
        let add128_path = risc0_root.join("bigint2/src/add128");
        self.copy_file(&src_path, &add128_path, "add128.blob");
```

We may now bootstrap the risc0 repository, including the new blob file, like this (you may need to adjust the path):

```
cargo bootstrap bigint2 --output=`$HOME`/risc0/
```

> To invoke the new addition program, we move over to the `risc0` repository. Here we will create a new module, akin to `bigint2/src/ec` or `bigint2/src/rsa`, containing a new `bigint2/src/add128/mod.rs` file. Its contents should look like this:

Reviewer: ```suggestion
## Invoking a bigint blob
To invoke the new addition program, we move over to the `risc0` repository. Here we will create a new module, akin to `bigint2/src/ec` or `bigint2/src/rsa`, containing a new `bigint2/src/add128/mod.rs` file. Its contents should look like this:
```

Reviewer: Mb "Invoking a bigint program" is better here?

Reviewer: or "Invoking a bibc blob?"

Reviewer 2: Personally I would go w/ invoking a bigint program (or precompile, maybe)
```suggestion
## Invoking a bigint program
To invoke the new addition program, we move over to the `risc0` repository. Here we will create a new module, akin to `bigint2/src/ec` or `bigint2/src/rsa`, containing a new `bigint2/src/add128/mod.rs` file. Its contents should look like this:
```
